pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -66,6 +66,59 @@ type DefaultPreemption struct {
 
 var _ framework.PostFilterPlugin = &DefaultPreemption{}
 
+type findCandidatesOpts struct {
+	candidateOffset       *int32
+	defaultPreemptionArgs *config.DefaultPreemptionArgs
+}
+
+// calculateNumCandidates returns the number of candidates the FindCandidates
+// method must produce from dry running based on the constraints given by
+// <minCandidateNodesPercentage> and <minCandidateNodesAbsolute>. The number of candidates
+// returned will never be greater than <numNodes>.
+func (fco *findCandidatesOpts) calculateNumCandidates(numNodes int32) int32 {
+	n := (numNodes * fco.defaultPreemptionArgs.MinCandidateNodesPercentage) / 100
+	if n < fco.defaultPreemptionArgs.MinCandidateNodesAbsolute {
+		n = fco.defaultPreemptionArgs.MinCandidateNodesAbsolute
+	}
+	if n > numNodes {
+		n = numNodes
+	}
+	return n
+}
+
+// updateAndReturnOffset updates its internal state returns an <offset> and
+// <numCandidates> that the caller must use to find candidates. A moving offset
+// is used as the starting index when dry running preemption so that the same
+// set of nodes don't get preempted every time. If <offset> + <numCandidates>
+// runs past the number of nodes, the caller is expected to wrap around the
+// node list.
+func (fco *findCandidatesOpts) updateAndReturnOffset(numNodes int32) (int32, int32) {
+	offset := *fco.candidateOffset
+	numCandidates := fco.calculateNumCandidates(numNodes)
+	if *fco.candidateOffset >= numNodes {
+		// This happens when the cluster is downscaled and candidateOffset is past
+		// the new, reduced number of nodes. Reset the counter and restart from the
+		// beginning.
+		*fco.candidateOffset = 0
+		offset = 0
+	} else {
+		*fco.candidateOffset = (*fco.candidateOffset + numCandidates) % numNodes
+	}
+	return offset, numCandidates
+}
+
+func withDefaultPreemptionArgs(args *config.DefaultPreemptionArgs) func(*findCandidatesOpts) {
```

Comment:
This function needs to be exported. Because of that, it shouldn't depend on DefaultPreemptionArgs. Just expose each parameter individually.

Reply:
Ah I see, done.